### PR TITLE
Update LabelModel config

### PR DESCRIPTION
**Pull request by bhancock8**
- Replace config dict with Config namedtuples (`LabelModelConfig` and `TrainConfig`)
- Delete recursive_merge_dict 🎉 
- Refactor train_model() -> fit(); log_train_every -> log_freq
- Move optimizers, lr_schedulers, and `merge_config()` to utils/ (since they're used by both Classifier and LabelModel)
- Added docstrings describing each attribute in the configs
- Move lr/l2 directly accessible in `TrainConfig`/`TrainerConfig` instead of nesting in `OptimizerConfig`

**Test plan:**
- Updated tests to use the new signatures
- ✔ `tox`

### CODE REVIEW

File: snorkel/classification/utils.py [hunk 1]
```diff
@@ -108,66 +107,6 @@ def merge_config(config: Config, config_updates: Dict[str, Any]) -> Config:
     return config._replace(**config_updates)
 
 
-def recursive_merge_dicts(
```
Comment (henryre) [inline snorkel/classification/utils.py]: 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏

File: snorkel/utils/optimizers.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    """Settings for Adamax optimizer."""
+
+    betas: Tuple[float, float] = (0.9, 0.999)
+    eps: float = 1e-8
+
+
+class OptimizerConfig(Config):
+    """Settings common to all optimizers."""
+
+    lr: float = 0.001
```
Comment (henryre) [inline snorkel/utils/optimizers.py]: Hmmmm this is a pretty critical value. Do we believe this is better than 0.01 for the label model? cc @paroma

Comment (henryre) [reply snorkel/utils/optimizers.py]: Can we surface lr and l2 one level higher? Just since they're particularly important

File: test/labeling/test_convergence.py [hunk 3]
```diff
@@ -82,7 +82,7 @@ def test_labeling_convergence(self) -> None:
 
         # Train LabelModel
         label_model = LabelModel(cardinality=self.cardinality, verbose=False)
-        label_model.train_model(L_train, lr=0.01, l2=0.0, n_epochs=100)
+        label_model.fit(L_train, lr=0.01, l2=0.0, n_epochs=100)
```
Comment (henryre) [inline test/labeling/test_convergence.py]: wait why can we do lr at the top level here?

File: snorkel/utils/optimizers.py (hunk 2 above)
Comment (paroma) [reply snorkel/utils/optimizers.py]: Default `n_epochs` changed to account for lower LR. 0.01 is better for LabelModel (as in, it works most of the time with the different matrices we've had across tests and tutorials). I don't have a strong opinion, we can explicitly change LR in the label model if needed

File: snorkel/labeling/model/label_model.py [hunk 4]
```diff
@@ -9,17 +9,69 @@
...
+
+    Parameters
+    ----------
+    n_epochs
+        The number of epochs to train (where each epoch is a single optimization step)
+    l2
+        Centered L2 regularization strength (int, float, or np.array)
+    optimizer
+        Which optimizer to use (one of ["sgd", "adam", "adamax"])
+    optimizer_config
```
Comment (paroma) [inline snorkel/labeling/model/label_model.py]: is lr in the config for sgd? maybe a pointer for that

File: snorkel/labeling/model/label_model.py [hunk 5]
```diff
@@ -416,7 +460,7 @@ def score(
         -------
         >>> L = np.array([[1, 1, -1], [0, 0, -1], [1, 1, -1]])
         >>> label_model = LabelModel(verbose=False)
-        >>> label_model.train_model(L)
+        >>> label_model.fit(L, seed=123)
```
Comment (paroma) [inline snorkel/labeling/model/label_model.py]: is the seed necessary?

File: snorkel/labeling/model/label_model.py [hunk 6]
```diff
@@ -376,7 +420,7 @@ def predict(
         -------
         >>> L = np.array([[0, 0, -1], [1, 1, -1], [0, 0, -1]])
         >>> label_model = LabelModel(verbose=False)
-        >>> label_model.train_model(L)
+        >>> label_model.fit(L, seed=123)
```
Comment (paroma) [inline snorkel/labeling/model/label_model.py]: same as below

File: test/labeling/model/test_label_model.py [hunk 7]
```diff
@@ -242,45 +244,97 @@ def test_model_loss(self):
...
 
-        label_model.train_model(L, n_epochs=10, lr=0.01, momentum=0.9)
+        label_model.fit(L, n_epochs=10)
         next_loss = label_model._loss_mu().item()
 
         self.assertLessEqual(next_loss, init_loss)
 
         with self.assertRaisesRegex(Exception, "Loss is NaN."):
-            label_model.train_model(L, n_epochs=10, lr=1e8)
+            label_model.fit(L, n_epochs=10, optimizer_config={"lr": 1e8})
```
Comment (paroma) [inline test/labeling/model/test_label_model.py]: we can initialize LabelModel again here so the lr doesn't have to passed in like this (we don't want to encourage this usage right?)

File: test/labeling/model/test_label_model.py [hunk 8]
```diff
@@ -299,7 +353,7 @@ def test_label_model(self) -> None:
 
         # Train LabelModel
         label_model = LabelModel(cardinality=self.cardinality, verbose=False)
-        label_model.train_model(L, lr=0.01, l2=0.0, n_epochs=100)
+        label_model.fit(L, n_epochs=200, optimizer_config={"lr": 0.01}, seed=123)
```
Comment (paroma) [inline test/labeling/model/test_label_model.py]: same as above

File: tutorials/workshop/Workshop_3_Generative_Model_Training.ipynb [hunk 9]
```diff
@@ -122,12 +122,12 @@
     "from snorkel.analysis.utils import convert_labels\n",
     "\n",
     "label_model = LabelModel(k=2, verbose=True, seed=123)\n",
-    "label_model.train_model(\n",
+    "label_model.fit(\n",
```
Comment (paroma) [inline tutorials/workshop/Workshop_3_Generative_Model_Training.ipynb]: Line above also need `cardinality=2` instead of `k=2`

File: snorkel/labeling/model/label_model.py (hunk 5 above)
Comment (bhancock8) [reply snorkel/labeling/model/label_model.py]: Not anymore. The original lr (0.01) and n_epochs (100) were restored, and we seem to get consistent convergence for these examples.

**Comment by bhancock8** [comment]
@paroma lr is now accessible top-level, which addresses most of your comments. Other suggested changes were made as well.
